dataproc/quickstart/quickstart.go

```diff
@@ -0,0 +1,221 @@
+	if err != nil {
+		log.Fatalf("error creating the cluster client: %s", err)
+	}
+
+	// Create the cluster config.
+	createReq := &dataprocpb.CreateClusterRequest{
+		ProjectId: projectID,
+		Region:    region,
+		Cluster: &dataprocpb.Cluster{
+			ProjectId:   projectID,
+			ClusterName: clusterName,
+			Config: &dataprocpb.ClusterConfig{
+				MasterConfig: &dataprocpb.InstanceGroupConfig{
+					NumInstances:   1,
+					MachineTypeUri: "n1-standard-1",
+				},
+				WorkerConfig: &dataprocpb.InstanceGroupConfig{
+					NumInstances:   2,
+					MachineTypeUri: "n1-standard-1",
+				},
+			},
+		},
+	}
+
+	// Create the cluster.
+	createOp, err := clusterClient.CreateCluster(ctx, createReq)
+	if err != nil {
+		log.Fatalf("error submitting the cluster creation request: %v", err)
+	}
+
+	createResp, err := createOp.Wait(ctx)
+	if err != nil {
+		log.Fatalf("error creating the cluster: %v", err)
+	}
+
+	// Output a success message.
+	fmt.Printf("Cluster created successfully: %q", createResp.ClusterName)
+
+	// Create the job client.
+	jobClient, err := dataproc.NewJobControllerClient(ctx, option.WithEndpoint(endpoint))
+
+	// Create the job config.
+	submitJobReq := &dataprocpb.SubmitJobRequest{
+		ProjectId: projectID,
+		Region:    region,
+		Job: &dataprocpb.Job{
+			Placement: &dataprocpb.JobPlacement{
+				ClusterName: clusterName,
+			},
+			TypeJob: &dataprocpb.Job_PysparkJob{
+				PysparkJob: &dataprocpb.PySparkJob{
+					MainPythonFileUri: jobFilePath,
+				},
+			},
+		},
+	}
+
+	submitJobResp, err := jobClient.SubmitJob(ctx, submitJobReq)
+	if err != nil {
+		fmt.Errorf("error submitting job: %v", err)
+	}
+
+	id := submitJobResp.Reference.JobId
+
+	fmt.Printf("Submitted job %q", id)
+
+	// These states all signify that a job has terminated, successfully or not.
+	terminalStates := map[dataprocpb.JobStatus_State]bool{
+		dataprocpb.JobStatus_ERROR:     true,
+		dataprocpb.JobStatus_CANCELLED: true,
+		dataprocpb.JobStatus_DONE:      true,
+	}
+
+	// We can create a timeout such that the job gets cancelled if not in a terminal state after a certain amount of time.
+	timeout := 5 * time.Minute
+	start := time.Now()
+
+	state := submitJobResp.Status.State
+	for !terminalStates[state] {
+		if time.Since(start) > timeout {
+			cancelReq := &dataprocpb.CancelJobRequest{
+				ProjectId: projectID,
+				Region:    region,
+				JobId:     id,
+			}
+
+			_, err := jobClient.CancelJob(ctx, cancelReq)
+			if err != nil {
+				fmt.Errorf("error cancelling job: %v", err)
+			}
+			fmt.Errorf("job %q timed out after %d minutes", id, int64(timeout.Minutes()))
+			break
+		}
+
+		getJobReq := &dataprocpb.GetJobRequest{
+			ProjectId: projectID,
+			Region:    region,
+			JobId:     id,
+		}
+		getJobResp, err := jobClient.GetJob(ctx, getJobReq)
+		if err != nil {
+			fmt.Errorf("error getting job %q with error: %v", id, err)
+			break
+		}
+		state = getJobResp.Status.State
+
+		time.Sleep(1 * time.Second)
```

Comment:
Any particular reason for 1 second? Also, add a comment saying why this sleeps?

Reply:
This is to prevent excess polling of the api. The value is fairly arbitrary but I've been using 1 second for all of the quickstarts